lesson22/index.js

```diff
@@ -0,0 +1,180 @@
+const tableWrapper = document.getElementById("js-table-wrapper");
+
+const createElementWithClassName = (type, className) => {
+  const element = document.createElement(type);
+  element.className = className;
+  return element;
+};
+
+const createLoading = () => {
+  const loading = createElementWithClassName("img", "loading");
+  loading.id = "js-loading";
+  loading.src = "./img/loading.gif";
+  return loading;
+};
+
+const removeLoading = () => document.getElementById("js-loading").remove();
+
+const createErrorMessage = (error) => {
+  const errorMessage = createElementWithClassName("p", "error-message");
+  errorMessage.textContent = error;
+  return errorMessage;
+};
+
+const fetchErrorHandling = async (response) => {
+  if (response.ok) {
+    return await response.json();
+  } else {
+    throw new Error("サーバーエラーが発生しました");
+  }
+};
+
+const getJsonOrError = async (url) => {
+  const response = await fetch(url);
+  const json = await fetchErrorHandling(response);
+  return json;
+};
+
+const getTableData = async () => {
+  return new Promise((resolve) => {
+    setTimeout(
+      () => resolve(getJsonOrError("https://myjson.dit.upm.es/api/bins/9me1")),
+      3000
+    );
+  });
+};
+
+const tryGetTableData = async () => {
+  tableWrapper.appendChild(createLoading());
+  try {
+    const tableData = await getTableData();
+    if (tableData.length === 0) {
+      tableWrapper.appendChild(createErrorMessage("データがありません"));
+      return;
+    }
+    return tableData;
+  } catch (e) {
+    tableWrapper.appendChild(createErrorMessage(e));
+  } finally {
+    removeLoading();
+  }
+};
+
+const init = async () => {
+  const data = await tryGetTableData();
+  data && tableWrapper.appendChild(createTable(data));
+  setClickInSortButton();
+};
+
+const createTable = (usersData) => {
+  const tableItems = {
+    userId: "ID",
+    name: "名前",
+    gender: "性別",
+    age: "年齢"
+  };
+  const table = document.createElement("table");
+  const columnKeys = Object.keys(tableItems);
+  const tableHead = createTableHead(tableItems);
+  const tableBody = createTableBody(usersData, columnKeys);
+  table.appendChild(tableHead).after(tableBody);
+  return table;
+};
+
+const createTableHead = (items) => {
+  const thead = document.createElement("thead");
+  const tr = document.createElement("tr");
+  Object.values(items).forEach((item) => {
+    const th = createElementWithClassName("th", "js-th");
+    th.textContent = item;
+    (item === "ID" || item === "年齢") && th.appendChild(createSortButton());
```

Comment:
IMO
Is it possible to display text dynamically instead of a fixed string in order to be able to support additional features?

Reply:
Thanks for the review :)

Changed the way th is generated to assign an id to each element.

In addSortButton, the element to which you want to add a sortButton is obtained by its id, and createSortButton is executed.

We also changed the process for sorting to be executed in initSortButton.

e3023ee